File: scripts/stream-server.js
const http = require("http");
const net = require("net");
const crypto = require("crypto");
const HTTP_PORT = 8080;
const RTSP_PORT = 8554;
const BOUNDARY = "ct_simulator_mjpeg_boundary";

const httpClients = new Set();
const rtspClients = new Set();
let latestFrameBuffer = null;
let lastFrameTime = Date.now();
const startTime = Date.now();

// Idle Heartbeat Timer: 画面停止時やバックグラウンド時にVLCが10秒でタイムアウト切断されるのを全自動防止
setInterval(() => {
    const now = Date.now();
    if (latestFrameBuffer && now - lastFrameTime >= 100) {
        if (rtspClients.size > 0) {
            broadcastRtspFrame(latestFrameBuffer);
        }
    }
}, 100);

// ===================================================
// 1. HTTP ストリーミングサーバー (Port 8080)
// ===================================================
const httpServer = http.createServer((req, res) => {
    res.setHeader("Access-Control-Allow-Origin", "*");
    res.setHeader("Access-Control-Allow-Methods", "GET, POST, OPTIONS");
    res.setHeader("Access-Control-Allow-Headers", "Content-Type");

    if (req.method === "OPTIONS") {
        res.writeHead(204);
        res.end();
        return;
    }

    // シミュレータからのフレーム受信 (POST /upload-frame)
    if (req.method === "POST" && req.url === "/upload-frame") {
        const chunks = [];
        req.on("data", (chunk) => {
            chunks.push(chunk);
        });
        req.on("end", () => {
            const buffer = Buffer.concat(chunks);
            if (buffer.length > 0) {
                lastFrameTime = Date.now();
                if (buffer[0] === 0xff && buffer[1] === 0xd8) {
                    // 生の JPEG バイナリデータの直接処理（超高速）
                    latestFrameBuffer = buffer;
                } else {
                    const body = buffer.toString("utf8");
                    if (body.startsWith("data:image/jpeg;base64,")) {
                        const base64Data = body.replace(/^data:image\/jpeg;base64,/, "");
                        latestFrameBuffer = Buffer.from(base64Data, "base64");
                    }
                }

                if (latestFrameBuffer) {
                    broadcastHttpMjpeg(latestFrameBuffer);
                    broadcastRtspFrame(latestFrameBuffer);
                }
            }
            res.writeHead(200, { "Content-Type": "application/json" });
            res.end(JSON.stringify({ success: true, httpViewers: httpClients.size, rtspViewers: rtspClients.size }));
        });
        return;
    }

    // ブラウザ / VLC 向け HTTP MJPEG エンドポイント
    if (req.method === "GET" && (req.url.startsWith("/live/") || req.url === "/")) {
        if (req.url === "/") {
            res.writeHead(200, { "Content-Type": "text/html; charset=utf-8" });
            res.end(`
                <!DOCTYPE html>
                <html>
                <head><title>CT Simulator Stream Server</title></head>
                <body style="font-family: sans-serif; background: #0f172a; color: #f8fafc; padding: 2rem;">
                    <h1>🎥 CT Simulator Streaming Server</h1>
                    <p><strong>HTTP Port:</strong> ${HTTP_PORT} | <strong>RTSP Port:</strong> ${RTSP_PORT}</p>
                    <p><strong>HTTP Viewers:</strong> ${httpClients.size} | <strong>RTSP Viewers:</strong> ${rtspClients.size}</p>
                    <hr style="border-color: #334155;" />
                    <h3>📡 Stream Endpoints:</h3>
                    <ul>
                        <li><strong>HTTP MJPEG (VLC/Browser):</strong> <a href="http://127.0.0.1:${HTTP_PORT}/live/ct-camera.mjpg" target="_blank" style="color: #38bdf8;">http://127.0.0.1:${HTTP_PORT}/live/ct-camera.mjpg</a></li>
                        <li><strong>RTSP Stream:</strong> <code style="color: #a7f3d0;">rtsp://127.0.0.1:${RTSP_PORT}/live/ct-camera.mp4</code></li>
                    </ul>
                    <div style="margin-top: 1rem;">
                        <img src="/live/ct-camera.mjpg" style="border: 2px solid #38bdf8; max-width: 640px; border-radius: 8px;" alt="Live Stream Preview" />
                    </div>
                </body>
                </html>
            `);
            return;
        }

        res.writeHead(200, {
            "Content-Type": `multipart/x-mixed-replace; boundary=${BOUNDARY}`,
            "Cache-Control": "no-cache, no-store, must-revalidate",
            Pragma: "no-cache",
            Expires: "0",
            Connection: "keep-alive",
        });

        httpClients.add(res);

        if (latestFrameBuffer) {
            sendHttpChunk(res, latestFrameBuffer);
        }

        req.on("close", () => {
            httpClients.delete(res);
        });
        return;
    }

    res.writeHead(404);
    res.end();
});

function sendHttpChunk(res, buffer) {
    try {
        res.write(`--${BOUNDARY}\r\n`);
        res.write("Content-Type: image/jpeg\r\n");
        res.write(`Content-Length: ${buffer.length}\r\n\r\n`);
        res.write(buffer);
        res.write("\r\n");
    } catch (e) {
        httpClients.delete(res);
    }
}

function broadcastHttpMjpeg(buffer) {
    for (const clientRes of httpClients) {
        sendHttpChunk(clientRes, buffer);
    }
}

// ===================================================
// 2. RTSP ストリーミングサーバー (Port 8554 / TCP Interleaved)
// ===================================================
const rtspServer = net.createServer((socket) => {
    let session = {
        id: Math.floor(Math.random() * 100000000),
        cseq: 1,
        channel: 0,
        isPlaying: false,
        socket: socket,
        framesSent: 0,
    };

    console.log(`[RTSP Log] Client connected from ${socket.remoteAddress}:${socket.remotePort}`);

    socket.on("data", (data) => {
        const msg = data.toString();
        const lines = msg.split("\r\n");
        const firstLine = lines[0] || "";
        const parts = firstLine.split(" ");
        const method = parts[0];

        const cseqMatch = msg.match(/CSeq:\s*(\d+)/i);
        if (cseqMatch) {
            session.cseq = cseqMatch[1];
        }

        if (method === "OPTIONS") {
            console.log(`[RTSP Log] Client CSeq=${session.cseq}: OPTIONS`);
            const resp =
                `RTSP/1.0 200 OK\r\n` +
                `CSeq: ${session.cseq}\r\n` +
                `Public: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN\r\n\r\n`;
            socket.write(resp);
            return;
        }

        if (method === "DESCRIBE") {
            console.log(`[RTSP Log] Client CSeq=${session.cseq}: DESCRIBE`);
            const sdp =
                `v=0\r\n` +
                `o=- ${Date.now()} 1 IN IP4 127.0.0.1\r\n` +
                `s=CT Simulator Virtual Camera\r\n` +
                `t=0 0\r\n` +
                `m=video 0 RTP/AVP 26\r\n` +
                `c=IN IP4 127.0.0.1\r\n` +
                `a=rtpmap:26 JPEG/90000\r\n` +
                `a=control:track0\r\n` +
                `a=framerate:30\r\n`;

            const resp =
                `RTSP/1.0 200 OK\r\n` +
                `CSeq: ${session.cseq}\r\n` +
                `Content-Type: application/sdp\r\n` +
                `Content-Length: ${Buffer.byteLength(sdp)}\r\n\r\n` +
                sdp;
            socket.write(resp);
            return;
        }

        if (method === "SETUP") {
            console.log(`[RTSP Log] Client CSeq=${session.cseq}: SETUP`);
            const transportMatch = msg.match(/Transport:\s*([^\r\n]+)/i);
            const reqTransport = transportMatch ? transportMatch[1] : "";

            let channel = 0;
            const interMatch = reqTransport.match(/interleaved=(\d+)(?:-(\d+))?/i);
            if (interMatch) {
                channel = parseInt(interMatch[1], 10);
            }
            session.channel = channel;

            let respTransport = reqTransport;
            if (!respTransport.includes("interleaved")) {
                respTransport = `RTP/AVP/TCP;unicast;interleaved=${channel}-${channel + 1}`;
            }

            const resp =
                `RTSP/1.0 200 OK\r\n` +
                `CSeq: ${session.cseq}\r\n` +
                `Transport: ${respTransport}\r\n` +
                `Session: ${session.id};timeout=60\r\n\r\n`;
            socket.write(resp);
            return;
        }

        if (method === "PLAY") {
            console.log(`[RTSP Log] Client CSeq=${session.cseq}: PLAY -> Streaming started`);
            session.isPlaying = true;
            rtspClients.add(session);

            const nowTime = Math.floor((Date.now() - startTime) * 90) >>> 0;
            const resp =
                `RTSP/1.0 200 OK\r\n` +
                `CSeq: ${session.cseq}\r\n` +
                `Session: ${session.id}\r\n` +
                `Range: npt=now-\r\n` +
                `RTP-Info: url=track0;seq=${rtpSeq & 0xffff};rtptime=${nowTime}\r\n\r\n`;
            socket.write(resp);

            if (latestFrameBuffer) {
                sendRtpFrame(session, latestFrameBuffer);
            }
            return;
        }

        if (method === "TEARDOWN") {
            console.log(`[RTSP Log] Client CSeq=${session.cseq}: TEARDOWN -> Stopped (Sent ${session.framesSent} frames)`);
            session.isPlaying = false;
            rtspClients.delete(session);
            const resp = `RTSP/1.0 200 OK\r\nCSeq: ${session.cseq}\r\n\r\n`;
            socket.write(resp);
            socket.end();
            return;
        }
    });

    socket.on("close", () => {
        console.log(`[RTSP Log] Client disconnected (Sent ${session.framesSent} frames)`);
        session.isPlaying = false;
        rtspClients.delete(session);
    });

    socket.on("error", (err) => {
        console.log(`[RTSP Log] Socket error: ${err.message}`);
        session.isPlaying = false;
        rtspClients.delete(session);
    });
});

let rtpSeq = 0;

// JPEGバイナリのSOF0/SOF2マーカーから動的に正確な画像解像度(Width / Height)を抽出する
function parseJpegDimensions(jpegBuffer) {
    let width = 640;
    let height = 480;
    try {
        let offset = 2; // SOI (0xff 0xd8) の次から
        while (offset < jpegBuffer.length - 8) {
            // パディングバイト (0xFF) のスキップ
            while (offset < jpegBuffer.length && jpegBuffer[offset] === 0xff) {
                offset++;
            }
            if (offset >= jpegBuffer.length) break;
            
            const marker = jpegBuffer[offset];
            offset++;
            
            if (marker === 0xd8 || marker === 0xd9 || (marker >= 0xd0 && marker <= 0xd7)) {
                continue;
            }
            
            if (offset + 2 > jpegBuffer.length) break;
            const len = jpegBuffer.readUInt16BE(offset);
            
            if (marker === 0xc0 || marker === 0xc2) {
                if (offset + 7 < jpegBuffer.length) {
                    height = jpegBuffer.readUInt16BE(offset + 3);
                    width = jpegBuffer.readUInt16BE(offset + 5);
                    break;
                }
            }
            
            offset += len;
        }
    } catch (e) {
        // パース失敗時はデフォルト維持
    }
    return { width, height };
}

function sendRtpFrame(session, jpegBuffer) {
    if (!session || !session.socket || session.socket.destroyed) return;
    if (!Buffer.isBuffer(jpegBuffer)) return;

    try {
        const dims = parseJpegDimensions(jpegBuffer);
        
        // JPEG全体をペイロードとして使用 (先頭のSOI 0xff 0xd8 を除く)
        // VLC等はこの内包されたDQT, DHT, SOF0を読み取ってRFCの標準テーブルを上書きしてくれる
        let rawJpeg = jpegBuffer;
        if (jpegBuffer.length > 2 && jpegBuffer[0] === 0xff && jpegBuffer[1] === 0xd8) {
            rawJpeg = jpegBuffer.slice(2);
        }

        const timestamp = Math.floor((Date.now() - startTime) * 90) >>> 0;
        const CHUNK_SIZE = 1400;
        const totalLen = rawJpeg.length;
        let offset = 0;

        while (offset < totalLen) {
            const isLast = (offset + CHUNK_SIZE >= totalLen);
            const chunkSize = Math.min(CHUNK_SIZE, totalLen - offset);
            const chunkData = rawJpeg.slice(offset, offset + chunkSize);

            const rtpHeaderLen = 12;
            const jpeghdrLen = 8;
            const payloadLen = rtpHeaderLen + jpeghdrLen + chunkSize;

            // 1. Interleaved Frame (TCP) Header (4 bytes)
            const tcpHeader = Buffer.alloc(4);
            tcpHeader[0] = 0x24; // '$'
            tcpHeader[1] = (session.channel !== undefined ? session.channel : 0) & 0xff; // Dynamic Channel
            tcpHeader.writeUInt16BE(payloadLen, 2);

            // 2. RTP Header (12 bytes)
            const rtpHeader = Buffer.alloc(rtpHeaderLen);
            rtpHeader[0] = 0x80; // V=2, P=0, X=0, CC=0
            rtpHeader[1] = (isLast ? 0x80 : 0x00) | 26; // M=1 (Marker: Frame Complete) for last chunk, PT=26 (JPEG)
            rtpHeader.writeUInt16BE(rtpSeq & 0xffff, 2);
            rtpHeader.writeUInt32BE(timestamp, 4);
            rtpHeader.writeUInt32BE(0x12345678, 8); // SSRC

            // 3. JPEG Header for RTP (8 bytes - RFC 2435)
            // タイプとQ値は適当な固定値 (VLCはペイロード内のマーカーを優先するため)
            const jpegHeader = Buffer.alloc(jpeghdrLen);
            jpegHeader[0] = 0x00; // Type Specific
            jpegHeader[1] = (offset >> 16) & 0xff; // Fragment Offset High
            jpegHeader[2] = (offset >> 8) & 0xff;  // Fragment Offset Mid
            jpegHeader[3] = offset & 0xff;         // Fragment Offset Low
            jpegHeader[4] = 0x01; // Type = 1 (4:2:0)
            jpegHeader[5] = 0x50; // Q = 80
            jpegHeader[6] = Math.min(255, Math.ceil(dims.width / 8));
            jpegHeader[7] = Math.min(255, Math.ceil(dims.height / 8));

            // 4. パケットの単一バッファ化一括送信
            const packet = Buffer.concat([tcpHeader, rtpHeader, jpegHeader, chunkData]);
            session.socket.write(packet);

            rtpSeq = (rtpSeq + 1) & 0xffff;
            offset += chunkSize;
        }

        session.framesSent++;
    } catch (e) {
        rtspClients.delete(session);
    }
}

function broadcastRtspFrame(buffer) {
    for (const session of rtspClients) {
        if (session.isPlaying) {
            sendRtpFrame(session, buffer);
        }
    }
}
// ===================================================
// 3. WebSocket Ingestion Server (Port 8080)
// ===================================================
httpServer.on("upgrade", (req, socket, head) => {
    if (req.headers["upgrade"] !== "websocket") {
        socket.end("HTTP/1.1 400 Bad Request");
        return;
    }
    const key = req.headers["sec-websocket-key"];
    const hash = crypto.createHash("sha1");
    hash.update(key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11");
    const acceptKey = hash.digest("base64");

    const headers = [
        "HTTP/1.1 101 Switching Protocols",
        "Upgrade: websocket",
        "Connection: Upgrade",
        `Sec-WebSocket-Accept: ${acceptKey}`
    ];
    socket.write(headers.join("\r\n") + "\r\n\r\n");

    let buffer = Buffer.alloc(0);
    let messageFrames = [];

    socket.on("data", (chunk) => {
        buffer = Buffer.concat([buffer, chunk]);
        while (buffer.length >= 2) {
            const fin = (buffer[0] & 0x80) === 0x80;
            const opcode = buffer[0] & 0x0f;
            const masked = (buffer[1] & 0x80) === 0x80;
            let payloadLen = buffer[1] & 0x7f;
            let offset = 2;

            if (payloadLen === 126) {
                if (buffer.length < 4) return;
                payloadLen = buffer.readUInt16BE(2);
                offset = 4;
            } else if (payloadLen === 127) {
                if (buffer.length < 10) return;
                payloadLen = Number(buffer.readBigUInt64BE(2));
                offset = 10;
            }

            let maskingKey;
            if (masked) {
                if (buffer.length < offset + 4) return;
                maskingKey = buffer.subarray(offset, offset + 4);
                offset += 4;
            }

            if (buffer.length < offset + payloadLen) return;

            const payload = buffer.subarray(offset, offset + payloadLen);
            if (masked) {
                for (let i = 0; i < payload.length; i++) {
                    payload[i] ^= maskingKey[i % 4];
                }
            }

            if (opcode === 0x8) {
                socket.end();
                return;
            } else if (opcode === 0x2 || opcode === 0x0) {
                messageFrames.push(payload);
                if (fin) {
                    const fullMessage = Buffer.concat(messageFrames);
                    messageFrames = [];
                    lastFrameTime = Date.now();
                    latestFrameBuffer = fullMessage;
                    broadcastHttpMjpeg(latestFrameBuffer);
                    broadcastRtspFrame(latestFrameBuffer);
                }
            }

            buffer = buffer.subarray(offset + payloadLen);
        }
    });
});

// サーバー起動
httpServer.listen(HTTP_PORT, () => {
    console.log(`===================================================`);
    console.log(`🎥 CT Simulator Multi-Protocol Stream Server Started`);
    console.log(`🌐 HTTP MJPEG (VLC/Browser): http://127.0.0.1:${HTTP_PORT}/live/ct-camera.mjpg`);
    console.log(`📡 RTSP Stream: rtsp://127.0.0.1:${RTSP_PORT}/live/ct-camera.mp4`);
    console.log(`===================================================`);
});

rtspServer.listen(RTSP_PORT);

File: scripts/stream_server.py
#!/usr/bin/env python3
"""
CT Simulator Multi-Protocol Stream Server (Refactored)
Provides HTTP MJPEG, RTSP, and WebSocket ingestion in a clean, non-blocking class architecture.
"""
import base64, hashlib, json, re, socket, struct, threading, time

HTTP_PORT, RTSP_PORT = 8080, 8554
BOUNDARY = "ct_simulator_mjpeg_boundary"

class CTStreamServer:
    def __init__(self):
        self.lock = threading.Lock()
        self.frame = None
        self.last_time = time.time()
        self.start_time = time.time()
        self.http_clients = set()
        self.rtsp_clients = {}
        self.rtp_seq = 0

    def parse_jpeg_dims(self, jpeg):
        w, h = 640, 480
        try:
            off, end = 2, len(jpeg)
            while off < end - 8:
                while off < end and jpeg[off] == 0xFF: off += 1
                if off >= end: break
                m, off = jpeg[off], off + 1
                if m in (0xD8, 0xD9) or 0xD0 <= m <= 0xD7: continue
                if off + 2 > end: break
                mlen = (jpeg[off] << 8) | jpeg[off + 1]
                if m in (0xC0, 0xC2) and off + 7 < end:
                    h, w = struct.unpack(">HH", jpeg[off+3:off+7])
                    break
                off += mlen
        except Exception: pass
        return w, h

    def broadcast(self, buffer):
        with self.lock:
            self.frame = buffer
            http_socks = list(self.http_clients)
            rtsp_sessions = list(self.rtsp_clients.values())

        # HTTP MJPEG Broadcast
        chunk = f"--{BOUNDARY}\r\nContent-Type: image/jpeg\r\nContent-Length: {len(buffer)}\r\n\r\n".encode() + buffer + b"\r\n"
        for sock in http_socks:
            try: sock.sendall(chunk)
            except Exception:
                with self.lock: self.http_clients.discard(sock)

        # RTSP RTP Broadcast
        if not rtsp_sessions: return
        w, h = self.parse_jpeg_dims(buffer)
        raw_jpeg = buffer[2:] if buffer.startswith(b"\xff\xd8") else buffer
        ts = int((time.time() - self.start_time) * 90000) & 0xFFFFFFFF
        chunk_sz, total = 1400, len(raw_jpeg)
        wb, hb = min(255, (w+7)//8), min(255, (h+7)//8)

        for sess in rtsp_sessions:
            if not sess.get("play") or sess.get("closed"): continue
            sock, ch, off = sess["socket"], sess["ch"], 0
            try:
                while off < total:
                    curr = min(chunk_sz, total - off)
                    data = raw_jpeg[off:off+curr]
                    is_last = (off + chunk_sz >= total)
                    
                    tcp = struct.pack(">BBH", 0x24, ch, 12 + 8 + curr)
                    rtp = struct.pack(">BBHII", 0x80, (0x80 if is_last else 0) | 26, self.rtp_seq & 0xFFFF, ts, 0x12345678)
                    jpg = struct.pack(">BBBBBBBB", 0, (off>>16)&0xFF, (off>>8)&0xFF, off&0xFF, 1, 0x50, wb, hb)
                    
                    sock.sendall(tcp + rtp + jpg + data)
                    self.rtp_seq = (self.rtp_seq + 1) & 0xFFFF
                    off += curr
                sess["frames"] += 1
            except Exception:
                sess["closed"] = True
                with self.lock: self.rtsp_clients.pop(sess["id"], None)

    def heartbeat(self):
        while True:
            time.sleep(0.1)
            if self.frame and time.time() - self.last_time >= 0.1 and self.rtsp_clients:
                self.broadcast(self.frame)

    def process_buffer(self, buf, is_ws=False):
        if not buf: return
        self.last_time = time.time()
        if is_ws:
            self.broadcast(buf)
            return
            
        idx = buf.find(b"\xff\xd8")
        if idx != -1:
            self.broadcast(buf[idx:])
            return
            
        try:
            txt = buf.decode("utf-8", "ignore")
            if "base64," in txt:
                self.broadcast(base64.b64decode(txt.split("base64,")[1].split('"')[0].split("'")[0].strip()))
        except Exception: pass

    def send_http(self, sock, code, ctype, body=None, extra=""):
        res = f"HTTP/1.1 {code}\r\nAccess-Control-Allow-Origin: *\r\nContent-Type: {ctype}\r\n"
        if body is not None:
            res += f"Content-Length: {len(body)}\r\n{extra}\r\n"
            sock.sendall(res.encode() + body)
        else:
            sock.sendall((res + extra + "\r\n").encode())

    def handle_http(self, sock):
        try:
            sock.settimeout(10.0)
            req = b""
            while b"\r\n\r\n" not in req:
                c = sock.recv(4096)
                if not c: return
                req += c
                
            hdr_str, body = req.split(b"\r\n\r\n", 1)
            hdr = hdr_str.decode("utf-8", "ignore")
            lines = hdr.split("\r\n")
            meth, path = (lines[0].split(" ") + ["", ""])[:2]

            if meth == "OPTIONS":
                self.send_http(sock, "204 No Content", "text/plain", b"", "Access-Control-Allow-Methods: GET, POST, OPTIONS\r\n")
                return

            if "Upgrade: websocket" in hdr or "upgrade: websocket" in hdr:
                self.handle_ws(sock, hdr, body)
                return

            if meth == "POST" and path == "/upload-frame":
                cl = int(next((l.split(":")[1].strip() for l in lines if l.lower().startswith("content-length:")), "0"))
                while len(body) < cl:
                    c = sock.recv(min(cl - len(body), 65536))
                    if not c: break
                    body += c
                self.process_buffer(body)
                with self.lock:
                    n_h, n_r = len(self.http_clients), len(self.rtsp_clients)
                self.send_http(sock, "200 OK", "application/json", json.dumps({"success": True, "httpViewers": n_h, "rtspViewers": n_r}).encode())
                return

            if meth == "GET" and path == "/":
                with self.lock:
                    n_h, n_r = len(self.http_clients), len(self.rtsp_clients)
                html = f"<html><body style='font-family:sans-serif;background:#0f172a;color:#f8fafc;padding:2rem;'><h1>[Stream Server] CT Simulator Stream Server (Optimized)</h1><p>HTTP: {HTTP_PORT} | RTSP: {RTSP_PORT}</p><p>Viewers - HTTP: {n_h} | RTSP: {n_r}</p><hr/><a href='/live/ct-camera.mjpg' style='color:#38bdf8'>MJPEG Stream</a><br/><br/><img src='/live/ct-camera.mjpg' style='max-width:640px;border:2px solid #38bdf8;border-radius:8px;'/></body></html>"
                self.send_http(sock, "200 OK", "text/html", html.encode())
                return

            if meth == "GET" and "/live/" in path:
                sock.settimeout(None)
                self.send_http(sock, "200 OK", f"multipart/x-mixed-replace; boundary={BOUNDARY}", None, "Cache-Control: no-cache\r\nConnection: keep-alive\r\n")
                with self.lock:
                    self.http_clients.add(sock)
                    if self.frame:
                        try: sock.sendall(f"--{BOUNDARY}\r\nContent-Type: image/jpeg\r\nContent-Length: {len(self.frame)}\r\n\r\n".encode() + self.frame + b"\r\n")
                        except Exception: self.http_clients.discard(sock)
                return

            self.send_http(sock, "404 Not Found", "text/plain", b"")
        except Exception: pass
        finally:
            if sock not in self.http_clients: sock.close()

    def handle_ws(self, sock, hdr, body):
        try:
            key = next((l.split(":")[1].strip() for l in hdr.split("\r\n") if l.lower().startswith("sec-websocket-key:")), "")
            accept = base64.b64encode(hashlib.sha1((key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11").encode()).digest()).decode()
            sock.sendall(f"HTTP/1.1 101 Switching Protocols\r\nUpgrade: websocket\r\nConnection: Upgrade\r\nSec-WebSocket-Accept: {accept}\r\n\r\n".encode())
            sock.settimeout(None)
            
            buf, frames = bytearray(body), bytearray()
            while True:
                c = sock.recv(4096)
                if not c: break
                buf.extend(c)
                while len(buf) >= 2:
                    fin, op, masked, plen, off = bool(buf[0]&0x80), buf[0]&0x0F, bool(buf[1]&0x80), buf[1]&0x7F, 2
                    if plen == 126: plen, off = struct.unpack(">H", buf[2:4])[0], 4
                    elif plen == 127: plen, off = struct.unpack(">Q", buf[2:10])[0], 10
                    if masked: mkey, off = buf[off:off+4], off+4
                    if len(buf) < off + plen: break
                    
                    pld = buf[off:off+plen]
                    if masked: pld = bytearray(b ^ mkey[i%4] for i, b in enumerate(pld))
                    
                    if op == 0x8: return
                    if op in (0x2, 0x0):
                        frames.extend(pld)
                        if fin:
                            self.process_buffer(bytes(frames), True)
                            frames.clear()
                    buf = buf[off+plen:]
        except Exception: pass
        finally: sock.close()

    def handle_rtsp(self, sock):
        sid = int(time.time()*1000)%100000000
        sess = {"id": sid, "cseq": "1", "ch": 0, "play": False, "socket": sock, "frames": 0, "closed": False}
        def reply(code, extra=""): sock.sendall(f"RTSP/1.0 {code}\r\nCSeq: {sess['cseq']}\r\n{extra}\r\n".encode())

        try:
            while not sess["closed"]:
                d = sock.recv(2048)
                if not d: break
                msg = d.decode("utf-8", "ignore")
                sess["cseq"] = next((m.group(1) for m in [re.search(r"CSeq:\s*(\d+)", msg, re.I)] if m), sess["cseq"])
                meth = msg.split(" ")[0]

                if meth == "OPTIONS": reply("200 OK", "Public: OPTIONS, DESCRIBE, SETUP, PLAY, TEARDOWN\r\n")
                elif meth == "DESCRIBE":
                    sdp = f"v=0\r\no=- {int(time.time())} 1 IN IP4 127.0.0.1\r\ns=CT Stream\r\nt=0 0\r\nm=video 0 RTP/AVP 26\r\nc=IN IP4 127.0.0.1\r\na=rtpmap:26 JPEG/90000\r\na=control:track0\r\na=framerate:30\r\n"
                    reply("200 OK", f"Content-Type: application/sdp\r\nContent-Length: {len(sdp)}\r\n\r\n{sdp}")
                elif meth == "SETUP":
                    trans = next((m.group(1) for m in [re.search(r"Transport:\s*([^\r\n]+)", msg, re.I)] if m), "")
                    sess["ch"] = int(next((m.group(1) for m in [re.search(r"interleaved=(\d+)", trans, re.I)] if m), 0))
                    if "interleaved" not in trans: trans = f"RTP/AVP/TCP;unicast;interleaved={sess['ch']}-{sess['ch']+1}"
                    reply("200 OK", f"Transport: {trans}\r\nSession: {sid};timeout=60\r\n")
                elif meth == "PLAY":
                    sess["play"] = True
                    with self.lock: self.rtsp_clients[sid] = sess
                    now_rtp = int((time.time() - self.start_time) * 90000) & 0xFFFFFFFF
                    reply("200 OK", f"Session: {sid}\r\nRange: npt=now-\r\nRTP-Info: url=track0;seq={self.rtp_seq & 0xffff};rtptime={now_rtp}\r\n")
                    if self.frame: self.broadcast(self.frame)
                elif meth == "TEARDOWN":
                    sess["play"] = False
                    with self.lock: self.rtsp_clients.pop(sid, None)
                    reply("200 OK")
                    break
        except Exception as e: print(f"[RTSP Exception]: {e}")
        finally:
            sess["closed"] = True
            with self.lock: self.rtsp_clients.pop(sid, None)
            sock.close()

    def start_server(self, port, handler):
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        s.bind(("0.0.0.0", port))
        s.listen(50)
        while True:
            try: threading.Thread(target=handler, args=(s.accept()[0],), daemon=True).start()
            except Exception: break

    def run(self):
        threading.Thread(target=self.heartbeat, daemon=True).start()
        threading.Thread(target=self.start_server, args=(HTTP_PORT, self.handle_http), daemon=True).start()
        threading.Thread(target=self.start_server, args=(RTSP_PORT, self.handle_rtsp), daemon=True).start()
        print(f"===================================================\n[Stream Server] CT Simulator Stream Server (Optimized)\n[HTTP MJPEG]: http://127.0.0.1:{HTTP_PORT}/live/ct-camera.mjpg\n[RTSP Stream]: rtsp://127.0.0.1:{RTSP_PORT}/live/ct-camera.mp4\n===================================================")
        try:
            while True: time.sleep(1)
        except KeyboardInterrupt: print("\nShutting down stream server...")

if __name__ == "__main__":
    CTStreamServer().run()
